---
title: "W3 practical: example ChIPseq processing"
output: html_document
---

```{r}
suppressPackageStartupMessages({
  library(AnnotationHub)
  library(Rsubread)      # read alignment
  library(rtracklayer)   # read/write genomic file formats
  library(Biostrings)    # handling sequences
  library(Rfastp)        # read QC/trimming
  library(epiwraps)
})
ah <- AnnotationHub()
# if on renku, use ah <- AnnotationHub(cache="ahcache")
```

We will perform the primary analysis of an example dataset, at each step demonstrating the use of **both a very standard pipeline, and a full R-based equivalent**. Pay attention to which chunks are R or bash code (see header)

# Downloading the raw reads

We download an example dataset which is (a subset of) a ChIP-seq experiment mapping the bindings of the factor ttk across the Kc167 Drosophila cell line.

We download the data:

```{r, eval=FALSE}
# Eval = FALSE : dont execute this the next time you run this, prefer to download in the code so you have a record of what you donwloaded and not just go on a website
# not important in this case, but if downloading large files on a slow connection
# we need to increase the download timeout:
options(timeout=3600)

dir.create("raw")
download.file("https://www.ethz-ins.org/content/ttk.fastq.gz", "raw/ttk.fastq.gz", mode="wb")
download.file("https://www.ethz-ins.org/content/input.fastq.gz", "raw/input.fastq.gz", mode="wb")
```

# Reads QC and trimming

## Using fastqc

For this we will use shell commands:

```{bash engine.opts='-l', eval=FALSE}
# note that depending on how your computer/server is set up, you might need to 
# specify where to find the software, -l is for using local configuration, e.g.:
export PATH="/opt/miniconda3/bin:$PATH"
mkdir -p raw_fastqc
fastqc -o raw_fastqc --threads 4 raw/*.fastq.gz
```

This will create an html report with QC plots for each file. We could optionally aggregate those using multiQC:

```{bash engine.opts='-l', eval=FALSE}
multiqc -o raw_fastqc raw_fastqc
```

### Trimming

```{bash, eval=FALSE}
mkdir -p trimmed
# this would be for a single sample:
trimmomatic SE -threads 4 -summary trimmed/ttk.stats -phred33 raw/ttk.fastq.gz trimmed/ttk.fastq.gz \
ILLUMINACLIP:TruSeq3-SE:2:15:4:4:true LEADING:20 TRAILING:20 SLIDINGWINDOW:4:15 MINLEN:25
# we do the same for the input:
trimmomatic SE -threads 4 -summary trimmed/input.stats -phred33 raw/input.fastq.gz trimmed/input.fastq.gz ILLUMINACLIP:TruSeq3-SE:2:15:4:4:true LEADING:20 TRAILING:20 SLIDINGWINDOW:4:15 MINLEN:25
```

## Using R and Rfastp

```{r}
# for each file call the function rfastp
dir.create("rfastp.trimmed")
qc <- lapply(c(ttk="raw/ttk.fastq.gz", input="raw/input.fastq.gz"), FUN=function(x){
  Rfastp::rfastp(x, thread=4, overrepresentationAnalysis=TRUE,
                 outputFastq=file.path("rfastp.trimmed/",gsub("\\.fastq\\.gz$","",basename(x))))
})
```

This will create an html report with before/after QC plots, as well as a trimmed and filtered fastq file. Furthermore, the R output object (equivalent to the .json stats also saved by the program) can be used to generate the QC plots inside a markdown, for example:

```{r}
# you can get the similar metrics as fastqc in the following package, here is an example:
Rfastp::curvePlot(qc$ttk, curve="content_curves")
```

# Alignment

## Using bowtie2

### Building a bowtie2 index

This we would need to run only once for this genome:

```{bash engine.opts='-l'}
# bowtie2-build [options] path/to/genome.fasta path/to/outputIndex
bowtie2-build --threads 4 BDGP6_genome/genome.fasta.gz BDGP6_genome/bowtie2
```

### Alignment

```{bash engine.opts='-l'}

# here -U single end, if paired end , use U1 and U1 kind of
mkdir -p aligned
(bowtie2 -p 4 -x BDGP6_genome/bowtie2 -U trimmed/ttk.fastq.gz) 2> aligned/ttk.bowtie2 |\
  samtools view -bS - | samtools sort -@4 -m 2G - > aligned/ttk.bam
samtools index aligned/ttk.bam
(bowtie2 -p 4 -x BDGP6_genome/bowtie2 -U trimmed/input.fastq.gz) 2> aligned/input.bowtie2 |\
  samtools view -bS - | samtools sort -@4 -m 2G - > aligned/input.bam
samtools index aligned/input.bam
```

## In R using Rsubread

### Building a genome index for mapping

This we have to do only once for a genome, and can then re-use across projects

```{r, eval=FALSE}
# we get the genome sequence from AnnotationHub
genome <- ah[["AH49674"]]
# we create a new directory that will contain the genome index
dir.create("BDGP6_genome")
# we write the genome sequence in fasta format
export(import.2bit(genome), "BDGP6_genome/genome.fasta.gz", compress=TRUE)
# we build a Rsubread index
Rsubread::buildindex("BDGP6_genome/rsubread", reference="BDGP6_genome/genome.fasta.gz")
```

### Alignment

```{r}
dir.create("aligned")
# readfile1: single end, paired ends readfile1 and readfile2 
align.stats <- Rsubread::align(index="BDGP6_genome/rsubread", type="dna",
                               readfile1=c("rfastp.trimmed/ttk_R1.fastq.gz", 
                                           "rfastp.trimmed/input_R1.fastq.gz"),
                               output_file=c("aligned/ttk.bam","aligned/input.bam"),
                               nthreads=6, sortReadsByCoordinates=TRUE)
align.stats
```

# Peak calling

## Using MACS

```{bash engine.opts='-l'}
mkdir -p peaks
macs3 callpeak --outdir peaks -n ttk --gsize dm -t aligned/ttk.bam -c aligned/input.bam
# macs2 works in the same way
```

We can then import the peaks in R:


P-value:(-log10pvalues) proba if you would getvthis peak if you signal comes from your input, corrected
score is the  pvalue scaled from 0 to 1000
peak: within the range of the peak waht is the highest location
signalvalue: 4 you have 4 times more reads that what you have in the control.
Note that MACS can also be called from within R, through the `MACSr` package (which wraps around macs3):

```{r}
# install MACSr, if not already done:
if(!require("MACSr",quietly=TRUE)) BiocManager::install("MACSr")

# /!\ here before an empty peaks directory should be created in ther working folder
library(MACSr)
callpeak(tfile="aligned/ttk.bam", cfile = "aligned/input.bam",
         gsize="dm", outdir = "peaks", name = "ttk")
# (will take a while the first time it's run, as it installs a new environment)
```
```{r}
# run after MACsr in r
peaks <- rtracklayer::import("peaks/ttk_peaks.narrowPeak")
peaks
```

## Having a loog at a peak

```{r}
head(peaks)

plotSignalTracks(list(ttk="aligned/ttk.bam", input="aligned/input.bam"), region=peaks[3], extend=5000)

# we're going to put both tracks on the same scale, and add the peaks coordinates:
plotSignalTracks(list(ttk="aligned/ttk.bam", peaks=peaks, input="aligned/input.bam"), region=peaks[3],
                 extend=5000, tracks.params=list(ylim=c(0,50)))
```

Note that an accumulation of reads at a given locus is not necessarily a binding event, but might also be an artifact. This is the main reason for having a input control. For instance, let's have a look at the following region:

```{r}
plotSignalTracks(c(ttk="aligned/ttk.bam", input="aligned/input.bam"), region="2R:342663-342814",
                 tracks.params=list(ylim=c(0,80)))
```
